src/bp/ui-studio/src/web/components/ContentForm/UploadWidget/index.tsx

```diff
@@ -0,0 +1,208 @@
+import style from '~/views/FlowBuilder/sidePanelTopics/form/style.scss'
+
+import localStyle from './style.scss'
+
+const UploadWidget: FC<any> = props => {
+  const uploadReducer = (state, action) => {
+    if (action.type === 'uploadStart') {
+      return {
+        ...state,
+        error: null,
+        uploading: true
+      }
+    } else if (action.type === 'deleteFile') {
+      props.onChange(null)
+      return {
+        ...state,
+        error: null,
+        uploading: false
+      }
+    } else if (action.type === 'uploadError') {
+      const { error } = action.data
+
+      return {
+        ...state,
+        error,
+        uploading: false
+      }
+    } else if (action.type === 'uploadSuccess') {
+      const { url } = action.data
+
+      props.onChange(url)
+      return {
+        ...state,
+        error: null,
+        uploading: false
+      }
+    } else if (action.type === 'enterUrlManually') {
+      const { enterUrlManually } = action.data
+      return {
+        ...state,
+        error: null,
+        enterUrlManually
+      }
+    } else if (action.type === 'updateUrl') {
+      const { url } = action.data
+
+      return {
+        ...state,
+        error: null,
+        url
+      }
+    } else if (action.type === 'saveUrl') {
+      const { url } = action.data
+
+      props.onChange(url)
+      return {
+        ...state,
+        error: null
+      }
+    } else {
+      throw new Error("That action type isn't supported.")
+    }
+  }
+
+  const [state, dispatch] = useReducer(uploadReducer, {
+    error: null,
+    uploading: false,
+    enterUrlManually: false,
+    url: null
+  })
+
+  const { error, enterUrlManually, url } = state
+
+  const deleteFile = () => {
+    dispatch({ type: 'deleteFile' })
+  }
+
+  const startUpload = async event => {
+    const data = new FormData()
+    data.append('file', event.target.files[0])
+
+    dispatch({ type: 'uploadStart' })
+    await axios
+      .post(`${window.BOT_API_PATH}/media`, data, { headers: { 'Content-Type': 'multipart/form-data' } })
+      .then(response => {
+        const { url } = response.data
+
+        dispatch({ type: 'uploadSuccess', data: { url } })
+      })
+      .catch(e => {
+        dispatch({ type: 'uploadError', data: { error: e.message } })
+      })
+  }
+
+  const handleToggleManually = () => {
+    dispatch({ type: 'enterUrlManually', data: { enterUrlManually: !enterUrlManually } })
+  }
+
+  const handleUrlChange = (value: string) => {
+    dispatch({ type: 'updateUrl', data: { url: value } })
+  }
+
+  const saveUrl = () => {
+    dispatch({ type: 'saveUrl', data: { url } })
+  }
+
+  const isUrlOrRelativePath = (str: string) => {
+    const re = /^(?:[a-z]+:)?\/\/|^\//i
+
+    return re.test(str)
+  }
+
+  const { $filter: filter, $subtype: subtype, type } = props.schema
+  if (type !== 'string' || subtype !== 'media') {
+    return null
+  }
+
+  const { value } = props
+
+  const re = /^\/api\/.*\/bots\/.*\/media\/.*/g
```

Comment:
Good point, I extracted the regex in a function called isBpUrl and moved it into common (url.ts). 

I might be wrong, but I think I can't use the function inside the utils.js file since its a vanilla JS file called inside a VM, right?

Reply:
you're right yes